tests/execution/test_node_grouping.py

```diff
@@ -0,0 +1,143 @@
+    node_grouper = GroupNodesAllAsOne()
+    nodes_grouped = node_grouper.group_nodes(list(fn_graph.nodes.values()))
+    assert len(nodes_grouped) == 1
+
+
+def test_group_by_level():
+    # No good reason you'd ever really want to group this way, but this helps test the grouping
+    # system
+    fn_graph = FunctionGraph(no_parallel, config={})
+    node_grouper = GroupNodesByLevel()
+    nodes_grouped = node_grouper.group_nodes(list(fn_graph.nodes.values()))
+    assert len(nodes_grouped) == 6  # Two are in the same group
+
+
+def test_group_nodes_by_repeatable_blocks():
+    fn_graph = FunctionGraph(parallel_linear_basic, config={})
+    node_grouper = GroupByRepeatableBlocks()
+    nodes_grouped_by_name = {
+        group.base_id: group for group in node_grouper.group_nodes(list(fn_graph.nodes.values()))
+    }
+    assert len(nodes_grouped_by_name["collect-steps"].nodes) == 1
+    assert len(nodes_grouped_by_name["expand-steps"].nodes) == 1
+    assert len(nodes_grouped_by_name["block-steps"].nodes) == 3
+    assert len(nodes_grouped_by_name["number_of_steps"].nodes) == 1
+    assert len(nodes_grouped_by_name["final"].nodes) == 1
+    assert nodes_grouped_by_name["final"].purpose == NodeGroupPurpose.EXECUTE_SINGLE
+    assert nodes_grouped_by_name["number_of_steps"].purpose == NodeGroupPurpose.EXECUTE_SINGLE
+    assert nodes_grouped_by_name["block-steps"].purpose == NodeGroupPurpose.EXECUTE_BLOCK
+    assert nodes_grouped_by_name["collect-steps"].purpose == NodeGroupPurpose.GATHER
+    assert nodes_grouped_by_name["block-steps"].spawning_task_base_id == "expand-steps"
+    assert nodes_grouped_by_name["collect-steps"].spawning_task_base_id == "expand-steps"
+
+
+def test_group_nodes_by_repeatable_blocks_complex():
+    fn_graph = FunctionGraph(parallel_complex, config={})
+    node_grouper = GroupByRepeatableBlocks()
+    nodes_grouped_by_name = {
+        group.base_id: group for group in node_grouper.group_nodes(list(fn_graph.nodes.values()))
+    }
+    assert len(nodes_grouped_by_name["collect-steps"].nodes) == 1
+    assert len(nodes_grouped_by_name["expand-steps"].nodes) == 1
+    assert len(nodes_grouped_by_name["block-steps"].nodes) == 6
+    assert nodes_grouped_by_name["number_of_steps"].purpose == NodeGroupPurpose.EXECUTE_SINGLE
+    assert nodes_grouped_by_name["block-steps"].purpose == NodeGroupPurpose.EXECUTE_BLOCK
+    assert nodes_grouped_by_name["collect-steps"].purpose == NodeGroupPurpose.GATHER
+    assert nodes_grouped_by_name["block-steps"].spawning_task_base_id == "expand-steps"
+    assert nodes_grouped_by_name["collect-steps"].spawning_task_base_id == "expand-steps"
+
+
+def test_create_task_plan():
+    fn_graph = FunctionGraph(parallel_linear_basic, config={})
+    node_grouper = GroupByRepeatableBlocks()
+    nodes_grouped = node_grouper.group_nodes(list(fn_graph.nodes.values()))
+    print_node_groups(nodes_grouped)
+    # task_plan = grouping.create_task_plan(nodes_grouped)
+    # pdb.set_trace()
```

Comment:
is there some evaluation to be performed? You could capture the output of the print and validate against that? `caplog` is a built in thing pytest has -- see the caching unit tests.